crates/oci-distribution/src/lib.rs

```diff
@@ -0,0 +1,174 @@
+        let challenge_opt = auth.get::<BearerChallenge>();
+        if challenge_opt.is_none() {
+            return Ok(());
+        }
+
+        let challenge = challenge_opt.as_ref().unwrap()[0].clone();
+        let realm = challenge.realm.unwrap();
+        let mut oauth = OauthClient::new(
+            "krustlet",
+            Url::parse(realm.as_str())?,
+            Url::parse(realm.as_str())?,
+        );
+        oauth.add_scope(challenge.scope.unwrap().as_str());
+        oauth.set_client_secret(secret.unwrap_or_else(|| "".to_owned()));
+
+        let token = oauth
+            .exchange_client_credentials()
+            .with_client(&cli)
+            .execute::<StandardToken>()
+            .await?;
+
+        self.token = Some(token);
+
+        Ok(())
+    }
+
+    pub async fn pull_manifest(&self, image_name: String) -> OciResult<String> {
+        // We unwrap right now because this try_from literally cannot fail.
+        let reference = Reference::try_from(image_name).unwrap();
+        let url = reference.to_v2_manifest_url();
+        let res = reqwest::get(&url).await?;
+        let status = res.status();
+        if res.status().is_client_error() {
+            // According to the OCI spec, we should see an error in the message body.
+            let err = res.json::<OciEnvelope>().await?;
+            // FIXME: This should not have to wrap the error.
+            return Err(anyhow::format_err!("{}", err.errors[0]));
+        } else if status.is_server_error() {
+            return Err(anyhow::format_err!("Server error at {}", url));
+        }
+        let text = res.text().await?;
+        Ok(text)
+    }
+}
+
+#[derive(Clone)]
+struct BearerChallenge {
+    pub realm: Option<String>,
+    pub service: Option<String>,
+    pub scope: Option<String>,
+}
+
+impl Challenge for BearerChallenge {
+    fn challenge_name() -> &'static str {
+        "Bearer"
+    }
+    fn from_raw(raw: RawChallenge) -> Option<Self> {
+        match raw {
+            RawChallenge::Token68(_) => None,
+            RawChallenge::Fields(mut map) => Some(BearerChallenge {
+                // NB (mpb): This follows the `remove` pattern from the existing
+                // implementations. Not sure why they do this, though.
+                realm: map.remove("realm"),
+                scope: map.remove("scope"),
+                service: map.remove("service"),
+            }),
+        }
+    }
+    fn into_raw(self) -> RawChallenge {
+        let mut map = ChallengeFields::new();
+        self.realm
+            .and_then(|realm| map.insert_static_quoting("realm", realm));
+        self.scope
+            .and_then(|item| map.insert_static_quoting("scope", item));
+        self.service
+            .and_then(|item| map.insert_static_quoting("service", item));
+        RawChallenge::Fields(map)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    #[tokio::test]
+    async fn test_version() {
+        let c = Client::default();
```

Comment:
Do we really want to make network requests in tests?

Reply:
Again, I would normally say no... but we're doing it elsewhere. Today we talked about setting up some integration tests, in which case I will move these. But for now, I would prefer to leave these so that we have some sanity check on a protocol that actually does not follow the published standard.

Reply:
+1 for keeping the tests until we have some integration testing in place.